src/include/hardware.h

```diff
@@ -73,6 +73,49 @@
 #define WT_WRITE_ONCE(v, val) WT_RELEASE_WRITE_WITH_BARRIER(v, val)
 #endif
 
+/*
+ * The below assembly implements the read-acquire semantic. Acquire semantics prevent memory
+ * reordering of the read-acquire with any load or store that follows it in program order.
+ *
+ * The if branches get removed at compile time as the sizeof instruction evaluates at compile time.
+ * Additionally ldapr results in a loss of type checking, to circumvent this we have an if (0) block
+ * which contains the direct assignment. This forces the compiler to type check.
+ *
+ * Depending on the size of the given type we choose the appropriate ldapr variant, additionally the
+ * W register variants are used if possible which map to the lower word of the associated X
+ * register.
+ *
+ * Finally the "Q" constraint is used for the given input operand, this instructs the compiler to
+ * generate offset free ldapr instructions. ldapr instructions, prior to version RCpc 3, don't
+ * support offsets.
+ *
+ * The flag HAVE_RCPC is determined by the build system, if this macro is removed in the future be
+ * sure to remove that part of the compilation.
+ */
+#ifdef HAVE_RCPC
+#ifndef TSAN_BUILD
+#define WT_ACQUIRE_READ(v, val)                                      \
+    do {                                                             \
+        if (0) {                                                     \
+            (v) = (val);                                             \
```

Comment:
Apologies I lost first version of this comment to webui usage error.

This should check that the values are assignment compatible. It does not verify that `sizeof(v)==sizeof(val)` which appears to be implied by the unit test.

For `sizeof(v) < sizeof(val)` as the register size is selected based on the sizeof(val) it looks like you'll get expected truncation.

Though surprising result will occur if `sizeof(v) > sizeof(val)` and both types are signed. As the instruction zero extends the loaded value, you'll get the following:

```
short val = -32768;
short *valp = &val;
long v = 0;
WR_ACQUIRE_READ(v, *valp);
// v == 32768 (sign loss)
```

From what I can find there is no sign extended version of `LDAPR` until  v8.4 (RCPC2?) so a possible solution is enforcing that the size of the source and destination are the same, one possible approach:
```
if ((sizeof(v)==sizeof(val))) {
  // instruction selection
} else {
  // force compilation error
}
```

Reply:
bf307a5f7